File: backend/server/routers/workflows.py
# ---------------------------------------------------------------------------
# system: ModelArchivist
# file: workflows.py
# purpose: REST interface for workflows
# ---------------------------------------------------------------------------

import backend.repository.repository as repo
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from backend.exception import ArcException
from backend.repository.tables import DeploymentStatus

router = APIRouter()


class WorkflowSearchCriteria(BaseModel):
    required_tags: list[str] = []
    forbidden_tags: list[str] = []
    name_prefix: str = ''


class WorkflowIds(BaseModel):
    ids: list[str]


class WorkflowTagUpdate(WorkflowIds):
    add: list[str]
    remove: list[str]


@router.get('/workflows')
async def get_workflows() -> list[dict]:
    workflows = repo.list_workflows(True)
    return workflows

@router.get('/workflows/{id}')
async def get_workflow(id: str) -> dict | None:
    try:
        return repo.get_workflow(id)
    except ArcException as e:
        if e.code == ArcException.Code.UNKNOWN_WORKFLOW:
            raise HTTPException(404, e.message)
        raise HTTPException(400, e.message)

@router.post('/workflows/search')
async def search_workflows(criteria: WorkflowSearchCriteria) -> list[dict]:
    return repo.list_workflows(True, criteria.model_dump())


@router.post('/workflows/bulk/tags')
async def update_workflow_tags(data: WorkflowTagUpdate) -> dict:
    return repo.update_workflow_tags(data.ids, data.add, data.remove)


@router.post('/workflows/bulk/synchronize')
async def synchronize_workflows(data: WorkflowIds, simulate: bool = True) -> dict:
    return repo.workflow_batch_operation(data.ids, 'synchronize', simulate)


@router.post('/workflows/bulk/move')
async def move_workflows(data: WorkflowIds, destination: DeploymentStatus,
                         simulate: bool = True) -> dict:
    return repo.workflow_batch_operation(data.ids, 'move', simulate, destination)


@router.put('/workflows/{id}')
async def update_workflow(id: str, changed_workflow: dict) -> dict:
    changed_workflow['id'] = id
    try:
        return repo.update_workflow(changed_workflow)
    except ArcException as error:
        if error.code == ArcException.Code.UNKNOWN_WORKFLOW:
            raise HTTPException(404, error.message)
        if error.code == ArcException.Code.READ_ONLY:
            raise HTTPException(403, error.message)
        raise HTTPException(400, error.message)


@router.post('/workflows/{id}/synchronize')
async def synchronize_workflow(id: str, simulate: bool = True) -> dict:
    return repo.synchronize_workflow(id, simulate)


@router.post('/workflows/{id}/move')
async def move_workflow(id: str, destination: DeploymentStatus,
                        simulate: bool = True) -> dict:
    return repo.move_workflow(id, destination, simulate)

File: frontend/src/lib/workflows.ts
/* ---------------------------------------------------------------------------
 * system: ModelArchivist
 * file: frontend/src/lib/workflows.ts
 * purpose: Workflow handling
 * ---------------------------------------------------------------------------*/

import { identity, toWorkflow, type Workflow, type WorkflowSummary } from '$lib/objects';
import { getUrl, parseResponse, type ApiResult } from '$lib/api';

export type WorkflowSearchCriteria = {
    required_tags: string[];
    forbidden_tags: string[];
    name_prefix: string;
};

export type WorkflowDestination = 'working' | 'archive';

export async function getWorkflows(): Promise<ApiResult<WorkflowSummary[]>> {
    const response = await fetch(getUrl('/workflows'));
    return await parseResponse(response, identity, 'getWorkflows');
}

export async function searchWorkflows(criteria: WorkflowSearchCriteria):
    Promise<ApiResult<WorkflowSummary[]>> {
    const response = await fetch(getUrl('/workflows/search'), {
        method: 'POST', headers: {'Content-Type': 'application/json'},
        body: JSON.stringify(criteria)
    });
    return await parseResponse(response, identity, 'searchWorkflows');
}

export async function getWorkflow(id: string): Promise<ApiResult<Workflow>> {
    const response = await fetch(getUrl(`/workflows/${id}`));
    return await parseResponse(response, toWorkflow, 'getWorkflow');
}

export async function updateWorkflow(workflow: Workflow): Promise<ApiResult<Workflow>> {
    const response = await fetch(getUrl(`/workflows/${workflow.id}`), {
        method: 'PUT', headers: {'Content-Type': 'application/json'},
        body: JSON.stringify(workflow)
    });
    return await parseResponse(response, toWorkflow, 'updateWorkflow');
}

async function workflowOperation(path: string): Promise<ApiResult<Record<string, any>>> {
    const response = await fetch(getUrl(path), {method: 'POST'});
    return await parseResponse(response, identity, 'workflowOperation');
}

export async function syncWorkflow(id: string) {
    return workflowOperation(`/workflows/${id}/synchronize?simulate=false`);
}

export async function moveWorkflow(id: string, destination: WorkflowDestination) {
    return workflowOperation(`/workflows/${id}/move?destination=${destination}&simulate=false`);
}

export async function updateWorkflowTags(ids: string[], add: string[], remove: string[]) {
    const response = await fetch(getUrl('/workflows/bulk/tags'), {
        method: 'POST', headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({ids, add, remove})
    });
    return await parseResponse<Workflow[]>(response,
        value => value.workflows.map(toWorkflow), 'updateWorkflowTags');
}

export async function syncWorkflows(ids: string[]) {
    const response = await fetch(getUrl('/workflows/bulk/synchronize?simulate=false'), {
        method: 'POST', headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({ids})
    });
    return await parseResponse<Record<string, any>>(response, identity, 'syncWorkflows');
}

export async function moveWorkflows(ids: string[], destination: WorkflowDestination) {
    const response = await fetch(getUrl(
        `/workflows/bulk/move?destination=${destination}&simulate=false`), {
        method: 'POST', headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({ids})
    });
    return await parseResponse<Record<string, any>>(response, identity, 'moveWorkflows');
}
